rtrlib/lib/vector.c

```diff
@@ -0,0 +1,154 @@
+
+
+#include "vector.h"
+
+#include <math.h>
+
+int aspa_vector_create(struct aspa_vector **vector_pointer)
+{
+	const size_t DEFAULT_INITIAL_SIZE = 128;
+
+	// allocation the chunk of memory
+	struct aspas_record *data_field =
+		(struct aspas_record *)malloc(sizeof(struct aspas_record) * DEFAULT_INITIAL_SIZE);
+
+	// malloc failed so returning an error
+	if (data_field == NULL) {
+		return -1;
+	}
+
+	struct aspa_vector *vector = (struct aspa_vector *)malloc(sizeof(struct aspa_vector));
+
+	if (vector == NULL) {
+		return -1;
+	}
+
+	vector->capacity = DEFAULT_INITIAL_SIZE;
+	vector->size = 0;
+	vector->data = data_field;
+
+	*vector_pointer = vector;
+
+	return 0;
+}
+
+int aspa_vector_delete(struct aspa_vector *vector)
+{
+	// checking if one of the two pointers is potentially invalid
+	if (vector == NULL || vector->data == NULL) {
+		return -1;
+	}
+
+	// freeing the data
+	free(vector->data);
+	// freeing the object itself
+	free(vector);
+
+	return 0;
+}
+
+int aspa_vector_reallocate(struct aspa_vector *vector)
+{
+	// by how many more elements this vector can hold after the relocation
+	const size_t SIZE_INCREASE_EXPONENTIAL = 2;
+
+	// allocation the new chunk of memory
+	struct aspas_record *new_data_field = (struct aspas_record *)malloc(
+		sizeof(struct aspas_record) * vector->capacity * SIZE_INCREASE_EXPONENTIAL);
+
+	// malloc failed so returning an error
+	if (new_data_field == NULL) {
+		return -1;
+	}
+
+	// copying the data from the old location to the new one
+	memcpy(new_data_field, vector->data, vector->capacity * sizeof(struct aspas_record));
+	vector->data = new_data_field;
+	vector->capacity = powl(vector->capacity, SIZE_INCREASE_EXPONENTIAL);
+
+	return 0;
+}
+
+void aspa_vector_sort(struct aspa_vector *vector, struct aspas_record record)
+{
+	size_t j = vector->size;
+
+	// moving data along
+	while (j > 0 && vector->data[j - 1].customer_as_number > record.customer_as_number) {
```

Comment:
some sort of explanatory comment would be appreciated here

Reply:
done